### PR TITLE
udev: check systemd is running and do not use cg_kill() if not

### LINKED ISSUE #11645: udev kills all processes on reload, when it thinks it is running under systemd (but it is not)
**systemd version the issue has been seen with**
> 240

**Used distribution**
 > Debian

**Expected behaviour you didn't see**
Invoking `udevadm trigger` should not kill unrelated processes

**Unexpected behaviour you saw**
Invoking `udevadm trigger` kills unrelated processes.

**Steps to reproduce the problem**

If udev is executed outside a systemd unit, udev sometimes thinks it has been started as a systemd unit and enables the killing spree, which ends up killing unrelated processes. This affects mostly users of other PID1 implementations.

It seems to me that this check is too simplistic:

https://github.com/systemd/systemd/blob/v240/src/udev/udevd.c#L1817-L1828

I wonder what is a better way to check this. Maybe using `$INVOCATION_ID` is a better way to answer the question: am I running as a systemd unit?

This is forwarded from [debian bug 918764](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=918764)

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #11645.

Not tested. cc @fsateler.

### CODE REVIEW

File: src/udev/udevd.c [hunk 1]
```diff
@@ -1798,7 +1798,7 @@ static int run(int argc, char *argv[]) {
 
         dev_setup(NULL, UID_INVALID, GID_INVALID);
 
-        if (getppid() == 1) {
+        if (getppid() == 1 && sd_booted() > 0) {
```
Comment (fsateler) [inline src/udev/udevd.c:1801]: `sd_booted` is documented to be able to return negative errno on failure. It currently doesn't, but I think if sd_booted should ever return an error, udev should not enable the killing logic. Better to leave stray processes than to kill unrelated stuff.

Comment (yuwata) [reply src/udev/udevd.c:1801]: `sd_booted()` does not return negative errno. I've prepared PR #11661 for updating the document.

Comment (poettering) [reply src/udev/udevd.c:1801]: hmm, i think we should fix sd_booted() instead to return an error if we can't figure out whether /run/systemd/system/ exists

Comment (yuwata) [reply src/udev/udevd.c:1801]: @poettering #11665.

**Comment by yuwata** [comment]
@fsateler OK. For safety, using `sd_id128_get_invocation()` seems to be good. `sd_booted()` just checks /run/systemd exists or not, and it can be easily mimicked.

PTAL.

**Comment by poettering** [comment]
I don't like the INVOCATION_ID check I must say. The concept is designed to be generic, i.e. not specific to systemd (which is why the env var isn't prefixed with SYSTEMD_). Hence it's not really suitable as a check whether systemd is around. This really should be an sd_is_booted() check, as the original version did it, as that does the check we really want to make here: are we running in a systemd context.

**Comment by fsateler** [comment]
@poettering `sd_booted` only checks that systemd is pid1. It does not check that it is launched as a systemd unit. As I mentioned in the issue, this can be triggered in a systemd system by invoking udev via start-stop-daemon or using setsid.

**Comment by poettering** [comment]
> @poettering `sd_booted` only checks that systemd is pid1. It does not check that it is launched as a systemd unit. As I mentioned in the issue, this can be triggered in a systemd system by invoking udev via start-stop-daemon or using setsid.

well, but is that a problem really? I mean, you have to jump through quite some hoops to trigger that. I mean, why would you start udev with s-s-d on a systemd system? I think we should guard against accidental borkage, and against usage in common scenarios outside of systemd. But the s-s-d thing appears to be neither to me?

**Comment by fsateler** [comment]
> well, but is that a problem really?

I would say so. Killing processes is a rather strong move. Thus, I'd say it should be done only when very very sure it's ok.

**Comment by linkfanel** [comment]
> well, but is that a problem really? I mean, you have to jump through quite some hoops to trigger that. I mean, why would you start udev with s-s-d on a systemd system? I think we should guard against accidental borkage, and against usage in common scenarios outside of systemd. But the s-s-d thing appears to be neither to me?

Really, could you please, please be a little more extra conservative, and a little less glib, about that mass-process-killing feature in your UID 0 daemon?

I think that yes, that is a problem, on several levels.

First, this is similar closing a bug as wontfix because it's too unlikely to be triggered. Now I understand and agree that in some cases it makes sense, like when fixes get so contrived that they lose meaning. I also understand the concept of not overburdening code and maintainers with portability over negligible environments - although I don't necessarily agree. On the other hand, some people (I've seen Linux kernel developers explicitly do this for example) believe that it is good, and important, to go the extra mile to fix those lesser issues - if only to set a good example of coding practices in a reference piece of software.

But most importantly, this bug here is not a theoretical case up to software philosophy debates, this is a real bug that has bitten people in obscure ways and wreaked havoc on their systems, for months on. And given that the consequence is killing innocent processes - usually all of them - I think this is something that you don't want to let slip any instance of, and absolutely want to fix correctly and thoroughly.

Second, you rightly point out that the two additional cases that were mentioned triggering the bug under systemd, are not common usage. However, these are only two known and identified cases; do we know that there aren't more unknown and unforeseen issues linked to this? Remember how you also initially wrongly supposed that there must have been something broken defining the systemd cgroup hierarchy, and thus the issue could be dismissed. Let's be prudent about how things could go wrong.

Third, as the result is killing innocent processes, this stops being about what systemd chooses to support, and rather becomes what it tolerates or allows to be run. It makes sense to question the relevance of start-stop-daemon under systemd, but if you go on with a decision that it should be excluded from the scope of this issue, this is in effect a policy that forbids using start-stop-daemon or setsid under systemd. Your init system should serve and help the user, not police and enforce through process killing which setups the user is prevented from putting in place. I would also point out the unfortunate implication that more than just declining to support alternative tools, this would in effect result with your init enforcing a policy to forbid and kill competing init software components. Let's agree that this is not something we want to do.

I agree that it is important to strictly limit process killing to identified udev spawns or cases where it is positively confirmed that the udev instance was launched with all the guarantees of a systemd unit.

**Comment by poettering** [comment]
Well, $INVOCATION_ID can be set by anyone, not just systemd, and in fact it was designed as a generic interface other init systems could implement too. Checking for $INVOCATION_ID is hence certainly not suitable as a check, because it covers too much.

We don't support systems that run udev via s-s-d or some script. All our deps are set up for systemd-udevd.service. Let's not kid ourselves that not using systemd to start udevd could over realistically work...

**Comment by keszybz** [comment]
I have to agree with @poettering here: `sd_booted()` seems more appropriate. The corner case of running `systemd` and *yet* starting udev in some strange way is too much of a corner case.

@linkfanel does the first version of this patch (with `sd_booted()`) fix your case?

**Comment by linkfanel** [comment]
The solution to this issue is not a choice between either sd_booted() or sd_id128_get_invocation(). For the reasons given by @poettering sd_id128_get_invocation() seems unsuitable, but sd_booted() is also insufficient.

As I suggested, my opinion is either to come up with a way - if possible - to track or find all udev descendant processes and kill only those; or to ensure that udev was launched with all the guarantees of a systemd unit: sd_booted() is an obvious start for that, but not enough.

Personally I don't run systemd, nor is it installed on my system. It's more or less willingly - for lack of choice - that I still run udev; it runs under sysvinit, through the stock Debian packaging init script that uses start-stop-daemon, despite being apparently unsupported by upstream. It seems to "realistically work" this way enough for me and Debian maintainers, despite "not using systemd to start it".

@keszybz So as for your question, yes, I could say that sd_booted() is enough to put me at a safe distance from this process killer and that whatever still happens on systemd systems is not my problem; except that this attitude doesn't help anyone, and also that I'm SCARED to see the udev package, over which I have no choice, plagued by critical regressions like this one, and to see that lead developers don't seem to think there is a problem to fix there, and that despite sd_booted() I might not ever be able to severe enough ties to remain unaffected by whatever their software model might entail.

**Comment by poettering** [comment]
> As I suggested, my opinion is either to come up with a way - if possible - to track or find all udev descendant processes and kill only those

That exists. It's called cgroups. And we use it for this already.

**Comment by poettering** [comment]
> except that this attitude doesn't help anyone, and also that I'm SCARED to see the udev package, over which I have no choice, plagued by critical regressions like this one, and to see that lead developers don't seem to think there is a problem to fix there, and that despite sd_booted() I might not ever be able to sever enough ties to remain unaffected by whatever their software model might entail.

You know, with posts like this you are not really helping your case.

**Comment by yuwata** [comment]
> Well, $INVOCATION_ID can be set by anyone, not just systemd, and in fact it was designed as a generic interface other init systems could implement too. Checking for $INVOCATION_ID is hence certainly not suitable as a check, because it covers too much.

I see. I will revert this to the original version.

> I have to agree with @poettering here: `sd_booted()` seems more appropriate. The corner case of running `systemd` and _yet_ starting udev in some strange way is too much of a corner case.

Yeah, indeed it is a corner case.

**Comment by yuwata** [comment]
Reverted.

**Comment by CameronNemo** [comment]
If only sd_booted() is used, this is a partial fix. Please don't close #11645 without handling the case of systemd-udevd being run under non-systemd on a systemd-using system.

File: src/udev/udevd.c [hunk 2]
```diff
@@ -29,6 +29,7 @@
 
 #include "sd-daemon.h"
 #include "sd-event.h"
+#include "sd-id128.h"
```
Comment (linkfanel) [inline src/udev/udevd.c]: Isn't this a stray include now?

**Comment by linkfanel** [comment]
> That exists. It's called cgroups. And we use it for this already.

Yes, that could be a way. However, correct me if I'm wrong, but what I understand is that udev itself doesn't set up the cgroups to track the processes it targets for killing; instead it assumes it was done by systemd but doesn't check for that properly; then it punishes the user for those failings. So when you say "we use it already", does "we" mean udev (where that responsibility should be but isn't), or systemd (introducing some troublesome and error-prone coupling), or do you make no difference between the two in your project vision?

> You know, with posts like this you are not really helping your case.

You know, with posts like yours you are not really helping your case. I'm getting less and less inclined to run your software. Happy, or shall we stoop lower next?

I'm giving you feedback on how this issue makes me feel. Some businesses find that feedback from upset people, negative reviews, and the like, constitute the _best_ PR, marketing and customer engagement opportunities as they're a great chance to turn feelings and impressions from negative to positive - even without necessarily having to act concretely. But I guess you find my feedback... unwelcome and unhelpful. Maybe it really is - it must be my own fault as a user if udev kills processes and I fail to help about that.

File: src/udev/udevd.c (hunk 2 above)
Comment (yuwata) [reply src/udev/udevd.c]: Thanks. Dropped.

**Pull request merged**